### ISSUE TITLE
Misleading error messages about lifetimes and borrowing

### ISSUE THREAD

**Issue by qiujiangkun** [rust-lang/rust#64002]
The error message is misleading that it led me to another approach through which I can't fix my program.
```log
error[E0597]: `line` does not live long enough
  --> src/main.rs:24:30
   |
24 |         let spt: Vec<&str> = line.split(char::is_whitespace).collect();
   |                              ^^^^ borrowed value does not live long enough
...
27 |             "#mem" => { rule.mem.push((spt[1], spt[2])) }
   |                         -------- borrow later used here
...
30 |     }
   |     - `line` dropped here while still borrowed

```
In fact the problem has nothing to do with borrowing but lifetime. As soon as I changed all my `&'a str` references to `String` and adjusted relevant codes properly, it worked.

[Here's the code](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=5b1c1fc2337c6ce54d9bffc2dc8c4879):
```rust
struct Rule<'a> {
    name: &'a str,
    mem: Vec<(&'a str, &'a str)>
}
impl<'a> Rule<'a> {
    fn new() -> Rule<'a> {
        Rule{
            name: "",
            mem: vec![]
        }
    }
}
use std::fs::File;
use std::io::{BufReader, BufRead};

fn read_rules(filename: &str) -> Vec<Rule> {
    let file = File::open(filename).unwrap();
    let reader = BufReader::new(file);
    let rules = vec![];
    let mut rule = Rule::new();
    for (index, line) in reader.lines().enumerate() {
        let line = line.unwrap();
        let spt: Vec<&str> = line.split(char::is_whitespace).collect();
        match spt[0] {
            "#def" => { rule.name = spt[1]; }
            "#mem" => { rule.mem.push((spt[1], spt[2])) }
            _ => { println!("{}", line) }
        }
    }
    return rules;
}
fn main() {
    read_rules("foo");
}
```

**Comment by csmoe** [comment]
> In fact the problem has nothing to do with borrowing but lifetime.
no, I think you have misunderstood the lifetime, when we talk about lifetime in Rust, we always default to the fact that borrowing happens there, or in short, no loan, no lifetime.

**Comment by qiujiangkun** [comment]
I see what a mistake I've made, as rust freshman. I tried to push a couple of strings with short time to a HashMap, which outlives my strings. Maybe it's a better way to tell user where he made a mistake like this:
```log
error[E0597]: `line` does not live long enough
  --> src/main.rs:24:30
   |
24 |         let spt: Vec<&str> = line.split(char::is_whitespace).collect();
   |                              ^^^^ borrowed value does not live long enough
...
27 |             "#mem" => { rule.mem.push((spt[1], spt[2])) }
   |                                       -------- borrow later used here
...
30 |     }
   |     - `line` dropped here while still borrowed
```

**Comment by csmoe** [comment]
You can ask for details about `E0597` from rustc with `rustc --explain E0597`, and I suggest you re-read chapters about lifetime in the rust book.

```rust
let mut rule = Rule::new(); 
for (index, line) in reader.lines().enumerate() {
        let line = line.unwrap();
        let spt: Vec<&str> = line.split(char::is_whitespace).collect(); 
                                // ^`split(&self, p)` borrows a local resource `line`(it will be dropped when a loop ends).
        match spt[0] {
            "#def" => { rule.name = spt[1]; }
            "#mem" => { rule.mem.push((spt[1], spt[2])) } 
              // you store two references to `line`  into `rule`(declared before this `for` loop),
              // when `line` is dropped at the end of a loop, this two references still live in rule,
              // thus they turn into dangling references as the original resources(line) dropped.
            _ => { println!("{}", line) }
        }
    }
```
